File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/dataholders/HolderInfo.kt
package com.jetbrains.signatureverifier.serialization.dataholders

import com.jetbrains.signatureverifier.serialization.toDLSequence
import com.jetbrains.signatureverifier.serialization.toPrimitiveDLSequence
import kotlinx.serialization.Serializable
import org.bouncycastle.asn1.ASN1Primitive
import org.bouncycastle.asn1.x509.Holder

@Serializable
data class HolderInfo(
  val baseCertificateId: IssuerSerialInfo?,
  val entityName: List<GeneralNameInfo>?,
  val objectDigestInfo: ObjectDigestInfo?,
  val version: Int
) : EncodableInfo {

  constructor(holder: Holder) : this(
    holder.baseCertificateID?.let { IssuerSerialInfo(it) },
    holder.entityName?.let { it.names.map { name -> GeneralNameInfo(name) } },
    holder.objectDigestInfo?.let { ObjectDigestInfo(it) },
    holder.version
  )

  override fun toPrimitive(): ASN1Primitive = when (version) {
    0 -> when (entityName) {
      null ->
        TaggedObjectInfo.getTaggedObject(
          true,
          0,
          baseCertificateId!!.toPrimitive()
        ).toASN1Primitive()

      else ->
        TaggedObjectInfo.getTaggedObject(
          true,
          1,
          entityName.toPrimitiveDLSequence()
        ).toASN1Primitive()
    }

    1 -> listOf(
      baseCertificateId?.let {
        TaggedObjectInfo.getTaggedObject(
          false,
          0,
          it.toPrimitive()
        )
      },
      entityName?.let {
        TaggedObjectInfo.getTaggedObject(
          false,
          1,
          it.toPrimitiveDLSequence()
        )
      },
      objectDigestInfo?.let {
        TaggedObjectInfo.getTaggedObject(
          false,
          2,
          it.toPrimitive()
        )
      },
    ).toDLSequence()

    else -> throw IllegalArgumentException("Unexpected version $version")
  }
}

File: net/JetBrains.Serialization/src/Dataholders/HolderInfo.cs
using Newtonsoft.Json;
using Org.BouncyCastle.Asn1;
using JetBrains.Serialization;
using Org.BouncyCastle.Asn1.X509;

[JsonObject(MemberSerialization.OptIn)]
public class HolderInfo : IEncodableInfo
{
  [JsonProperty("BaseCertificateId")] private IssuerSerialInfo? _baseCertificateId;
  [JsonProperty("EntityName")] private List<GeneralNameInfo>? _entityName;
  [JsonProperty("ObjectDigestInfo")] private ObjectDigestInfo? _objectDigestInfo;
  [JsonProperty("Version")] private int _version;

  [JsonConstructor]
  public HolderInfo(
    IssuerSerialInfo? baseCertificateId,
    List<GeneralNameInfo>? entityName,
    ObjectDigestInfo? objectDigestInfo,
    int version
  )
  {
    _baseCertificateId = baseCertificateId;
    _entityName = entityName;
    _objectDigestInfo = objectDigestInfo;
    _version = version;
  }

  public HolderInfo(Holder holder)
  {
    _baseCertificateId = holder.BaseCertificateID != null ? new IssuerSerialInfo(holder.BaseCertificateID) : null;
    _entityName = holder.EntityName?.GetNames().Select(name => new GeneralNameInfo(name)).ToList();
    _objectDigestInfo = holder.ObjectDigestInfo != null ? new ObjectDigestInfo(holder.ObjectDigestInfo) : null;
    _version = holder.Version;
  }

  public Asn1Encodable ToPrimitive()
  {
    switch (_version)
    {
      case 0:
        if (_entityName == null)
        {
          return TaggedObjectInfo.GetTaggedObject(true, 0,
            _baseCertificateId?.ToPrimitive() ?? throw new InvalidOperationException());
        }

        return TaggedObjectInfo.GetTaggedObject(true, 1, _entityName.ToPrimitiveDerSequence());
      case 1:
        var items = new List<Asn1Encodable>();
        if (_baseCertificateId != null)
        {
          items.Add(TaggedObjectInfo.GetTaggedObject(false, 0, _baseCertificateId.ToPrimitive()));
        }

        if (_entityName != null)
        {
          items.Add(TaggedObjectInfo.GetTaggedObject(false, 1, _entityName.ToPrimitiveDerSequence()));
        }

        if (_objectDigestInfo != null)
        {
          items.Add(TaggedObjectInfo.GetTaggedObject(false, 2, _objectDigestInfo.ToPrimitive()));
        }

        return items.ToDerSequence();
      default:
        throw new ArgumentException($"Unexpected version {_version}");
    }
  }
}